### PR TITLE
fabtests/multi_ep, prov/util, prov/verbs: separate EP resources and reenable multi EP testing

### PR DESCRIPTION
**Pull request by aingerson**
The multi_ep test has been disabled from runfabtests.sh because most providers
have issues with the test.

This changes the structure of the test so that each EP has completely separate
tx and rx resources (versus sharing CQs). Most use cases with multiple EPs use
separate resources for each EP so this not only avoids issues with providers
who currently have a limitation with shared resources but also allows reenabling
of the test in runfabtests.sh by default.

The test keeps the shared AV and shared EQ since these do not have known issues
and are commonly shared between EPs in common practice.

Signed-off-by: Alexia Ingerson <alexia.ingerson@intel.com>

### CODE REVIEW

**Comment by aingerson** [comment]
@shefty There are probably 3 patches worth and I squashed them since they are all very closely related but let me know if you want me to separate.

File: fabtests/functional/multi_ep.c [hunk 1]
```diff
@@ -80,16 +89,35 @@ static int alloc_multi_ep_res()
...
 	int i;
 
-	for (i = 0; i < num_eps; i++)
+	FT_CLOSE_FID(data_mr);
+	for (i = 0; i < num_eps; i++) {
 		FT_CLOSE_FID(eps[i]);
+		FT_CLOSE_FID(txcqs[i]);
+		FT_CLOSE_FID(rxcqs[i]);
+		FT_CLOSE_FID(txcntrs[i]);
+		FT_CLOSE_FID(rxcntrs[i]);
```
Comment (j-xiong) [inline fabtests/functional/multi_ep.c:114]: Those arrays could be NULL.

File: fabtests/functional/multi_ep.c [hunk 2]
```diff
@@ -99,93 +127,139 @@ static void free_ep_res()
...
+{
+	int ret;
+
+	do {
+		ret = fi_recv(eps[idx], recv_bufs[idx], opts.transfer_size,
+			      data_desc, FI_ADDR_UNSPEC, &recv_ctx[idx]);
+		if (ret == -FI_EAGAIN)
+			(void) fi_cq_read(rxcqs[idx], NULL, 0);
+
+	} while (ret == FI_EAGAIN);
```
Comment (j-xiong) [inline fabtests/functional/multi_ep.c:140]: -FI_EAGAIN

**Comment by j-xiong** [comment]
bot:aws:retest

**Comment by j-xiong** [comment]
multi EP test failed for tcp:
```
12:00:59  - name:   fi_multi_ep -e rdm -v -p "tcp"
12:00:59    timestamp: 20220408-190059+0000
12:00:59    result: Fail
12:00:59    time:   2
12:00:59    server_cmd: /home/cstbuild/ofi-Install/libfabric/ofi_libfabric/PR-7629/4/reg/bin/fi_multi_ep -e rdm -v -p "tcp"   -s ci5
12:00:59    server_stdout: |
12:00:59      fi_send(): functional/multi_ep.c:173, ret=-97 (Address family not supported by protocol)
12:00:59      Creating 3 EPs
12:00:59      Send to all 3 remote EPs
12:00:59    client_cmd: /home/cstbuild/ofi-Install/libfabric/ofi_libfabric/PR-7629/4/reg/bin/fi_multi_ep -e rdm -v -p "tcp"   -s ci6 ci5
12:00:59    client_stdout: |
12:00:59      fi_send(): functional/multi_ep.c:173, ret=-22 (Invalid argument)
12:00:59      Creating 3 EPs
12:00:59      Send to all 3 remote EPs
```

**Comment by shefty** [comment]
We debugged the tcp problem.  There are problems in both the test and underlying fi_getinfo path.

**Comment by aingerson** [comment]
@j-xiong @shefty Sorry for the delay. Took a while to pinpoint various issues. This should pass now I think

**Comment by aingerson** [comment]
Well it passed fi_multi_ep but failed all the fi_getinfo tests... investigating

**Comment by aingerson** [comment]
@shefty I removed one of the hints format check because it was triggering a verbs UD failure and I don't think it's really needed since the issue we were trying to resolve had to do with the IP path which is fixes by the interface query check. Let me know if you want me to add the other one back in/debug verbs UD
```
diff --git a/prov/util/src/util_attr.c b/prov/util/src/util_attr.c
index d5c2e4842..3efba2a0c 100644
--- a/prov/util/src/util_attr.c
+++ b/prov/util/src/util_attr.c
@@ -1274,6 +1274,8 @@ void ofi_alter_info(struct fi_info *info, const struct fi_info *hints,
                /* This should stay before call to fi_alter_domain_attr as
                 * the checks depend on unmodified provider mr_mode attr */
                info->caps = ofi_get_info_caps(info, hints, api_version);
+               if (hints && hints->addr_format != FI_FORMAT_UNSPEC)
+                       info->addr_format = hints->addr_format;
 
                if ((info->domain_attr->mr_mode & FI_MR_LOCAL) &&
                    (FI_VERSION_LT(api_version, FI_VERSION(1, 5)) ||
```

**Comment by aingerson** [comment]
Updated with verbs format fixes

**Review by shefty — APPROVED** [review]
Waiting for CI, but changes look good.  I added v1.15.x label, so we can pick up the fixes in this PR.

**Comment by aingerson** [comment]
@shefty changes failed verbs;ofi_rxd because it returned ENODATA and runfabtests is run in strict mode which treats it as a failure. working on fixing it

**Review by shefty — APPROVED** [review]
det funker

**Comment by aingerson** [comment]
Looks like it finally passed. Had to add another patch to allow tests to request rdm/dgram for the test which lets us still test the udp (no rxd) path while skipping verbs;ofi_rxd which will return ENODATA.

**Pull request merged**